### PR TITLE
Refactor KestrelTrace and add to the ServiceContext. Close aspnet/KestrelHttpServer#141

### PR DESCRIPTION
**Pull request by drussilla**
- Remove singleton implementation from KestrelTrace
- Inject KestrelTrace instance through ServiceContext
- Update Sample application to show LoggerFactory possibilities
- Inject IKestrelTrace in to SocketOutput

### CODE REVIEW

File: samples/SampleApp/project.json [hunk 1]
```diff
@@ -1,7 +1,9 @@
 {
     "version": "1.0.0-*",
     "dependencies": {
-        "Microsoft.AspNet.Server.Kestrel": "1.0.0-*"
+        "Microsoft.AspNet.Server.Kestrel": "1.0.0-*",
+        "Microsoft.Framework.Logging.Abstractions": "1.0.0-*",
```
Comment (davidfowl) [inline samples/SampleApp/project.json]: You can remove this.

File: samples/SampleApp/Startup.cs [hunk 2]
```diff
@@ -4,13 +4,16 @@
...
 namespace SampleApp
 {
     public class Startup
     {
-        public void Configure(IApplicationBuilder app)
+        public void Configure(IApplicationBuilder app, ILoggerFactory loggerFactory)
         {
+            loggerFactory.MinimumLevel = LogLevel.Debug;
+            loggerFactory.AddConsole(LogLevel.Debug);
             app.Run(context =>
```
Comment (davidfowl) [inline samples/SampleApp/Startup.cs]: nit: Space between these lines.

File: test/Microsoft.AspNet.Server.KestrelTests/TestLogger.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+{
+    public class TestLogger : ILogger
+    {
+        public void Log(LogLevel logLevel, int eventId, object state, Exception exception, Func<object, Exception, string> formatter)
+        {
+        }
+
+        public bool IsEnabled(LogLevel logLevel)
+        {
+            return true;
```
Comment (davidfowl) [inline test/Microsoft.AspNet.Server.KestrelTests/TestLogger.cs]: false?

**Comment by drussilla** [comment]
@davidfowl Thanks, fixed in latest commit

**Comment by davidfowl** [comment]
@lodejard @halter73 We need to do a couple of things here:
- Sprinkle IsEnabled in a bunch of places so we don't destroy performance.
- Measure the performance in general after this change

**Comment by halter73** [comment]
Thanks again for doing this work! This gets a :shipit: from me.

I'm going to get @lodejard to comment on this and merge it asap.

File: src/Microsoft.AspNet.Server.Kestrel/project.json [hunk 4]
```diff
@@ -11,7 +11,8 @@
         "Microsoft.StandardsPolice": {
             "version": "1.0.0-*",
             "type": "build"
-        }
+        },
+        "Microsoft.Framework.Logging.Abstractions": "1.0.0-*"
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/project.json]: Make this alphabetical.

**Comment by halter73** [comment]
OK. I actually just noticed you no longer have the `Trace.WriteLine` replaced with `LogErrors` like you did in the previous PR. Can we bring that back?

**Comment by drussilla** [comment]
@halter73 Done. Thanks!

File: src/Microsoft.AspNet.Server.Kestrel/Http/Listener.cs [hunk 5]
```diff
@@ -50,11 +50,11 @@ protected Listener(ServiceContext serviceContext) : base(serviceContext)
         /// </summary>
         protected abstract UvStreamHandle CreateListenSocket(string host, int port);
 
-        protected static void ConnectionCallback(UvStreamHandle stream, int status, Exception error, object state)
+        protected void ConnectionCallback(UvStreamHandle stream, int status, Exception error, object state)
         {
             if (error != null)
             {
-                Trace.WriteLine("Listener.ConnectionCallback " + error.ToString());
+                Log.LogError("Listener.ConnectionCallback ", error);
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Http/Listener.cs]: Make this static again and change this to:

    ((Listener)state).Log.LogError("Listener.ConnectionCallback ", error);

You could make a local listener var and use it in the else block too if you want.

File: src/Microsoft.AspNet.Server.Kestrel/Networking/UvConnectRequest.cs [hunk 6]
```diff
@@ -40,7 +43,7 @@ public void Init(UvLoopHandle loop)
             Libuv.pipe_connect(this, pipe, name, _uv_connect_cb);
         }
 
-        private static void UvConnectCb(IntPtr ptr, int status)
+        private  static void UvConnectCb(IntPtr ptr, int status)
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Networking/UvConnectRequest.cs]: Nit: remove extra space.

**Comment by drussilla** [comment]
@halter73 Done, btw, should I make "done" comments with your name to let you know that I have fixed code or you will get automatic notification from github? Because I think it might be a bit annoing for you to get a lot of notifications.

**Comment by halter73** [comment]
"Done" comments shouldn't be necessary, but they don't hurt. I'm going to try to merge this in today :+1:

**Pull request merged**